flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractionUtils.java

```diff
@@ -0,0 +1,167 @@
+
+		public Type getReturnType() {
+			return returnType;
+		}
+
+		public String getName() {
+			return name;
+		}
+
+		public boolean executablesEquals(Method m) {
+			return executable.equals(m);
+		}
+
+		public boolean executablesEquals(Constructor<?> c) {
+			return executable.equals(c);
+		}
+	}
+
+	public static LambdaExecutable checkAndExtractLambda(Function function) {
+		try {
+			// get serialized lambda
+			Object serializedLambda = null;
+			for (Class<?> clazz = function.getClass(); clazz != null; clazz = clazz.getSuperclass()) {
+				try {
+					Method replaceMethod = clazz.getDeclaredMethod("writeReplace");
+					replaceMethod.setAccessible(true);
+					Object serialVersion = replaceMethod.invoke(function);
+
+					// check if class is a lambda function
+					if (serialVersion.getClass().getName().equals("java.lang.invoke.SerializedLambda")) {
+
+						// check if SerializedLambda class is present
+						try {
+							Class.forName("java.lang.invoke.SerializedLambda");
+						}
+						catch (Exception e) {
+							throw new UnsupportedOperationException("User code tries to use lambdas, but framework is running with a Java version < 8");
+						}
+						serializedLambda = serialVersion;
+						break;
+					}
+				}
+				catch (NoSuchMethodException e) {
+					// thrown if the method is not there. fall through the loop
+				}
+			}
+
+			// not a lambda method -> return null
+			if (serializedLambda == null) {
+				return null;
+			}
+
+			// find lambda method
+			Method implClassMethod = serializedLambda.getClass().getDeclaredMethod("getImplClass");
+			Method implMethodNameMethod = serializedLambda.getClass().getDeclaredMethod("getImplMethodName");
+			Method implMethodSig = serializedLambda.getClass().getDeclaredMethod("getImplMethodSignature");
+
+			String className = (String) implClassMethod.invoke(serializedLambda);
+			String methodName = (String) implMethodNameMethod.invoke(serializedLambda);
+			String methodSig = (String) implMethodSig.invoke(serializedLambda);
+
+			Class<?> implClass = Class.forName(className.replace('/', '.'), true, Thread.currentThread().getContextClassLoader());
+
+			// find constructor
+			if (methodName.equals("<init>")) {
+				Constructor<?>[] constructors = implClass.getDeclaredConstructors();
+				for (Constructor<?> constructor : constructors) {
+					if(getConstructorDescriptor(constructor).equals(methodSig)) {
+						return new LambdaExecutable(constructor);
+					}
+				}
+			}
+			// find method
+			else {
+				List<Method> methods = getAllDeclaredMethods(implClass);
+				for (Method method : methods) {
+					if(method.getName().equals(methodName) && getMethodDescriptor(method).equals(methodSig)) {
+						return new LambdaExecutable(method);
+					}
+				}
+			}
+			throw new Exception("No lambda method found.");
+		}
+		catch (Exception e) {
+			throw new RuntimeException("Could not extract lambda method out of function: " + e.getClass().getSimpleName() + " - " + e.getMessage(), e);
```

Comment:
RuntimeException is usually used for programming errors (wrong arguments, null divisions, ...).
Something that is expected to fail sometimes beyond the programmers control (I/O, code analysis, etc) should use checked exceptions.